### PR TITLE
Environment variables priority

### PR DESCRIPTION
**Pull request by ulyssessouza**
**What I did**
This PR is meant to define the precedence of the environment variables evaluation.

For the environment variables to be **available in the container**:
1. Command Line (`docker compose run --env <KEY[=VAL]>` https://docs.docker.com/engine/reference/commandline/compose_run/#options)
2. Compose File (`service::environment` section: https://docs.docker.com/compose/compose-file/#environment)
3. Compose File (`service::env_file` section file: https://docs.docker.com/compose/compose-file/#env_file)
4. Container Image `ENV` directive (https://docs.docker.com/engine/reference/builder/#env)

---

For the environment variables available for **`docker compose`'s runtime**:
1. Command Line parent command option `docker compose --env-file <FILE>` (or `.env` in the root of the project by default)
2. OS Environment Variables

Obs: Note that the variables available for runtime will not be available in the container unless it's imported by one of the methods on the list above

**Related issue**
Resolves https://github.com/docker/compose/issues/9521
Resolves https://github.com/docker/compose/issues/9638
Resolves https://github.com/docker/compose/issues/9608
Resolves https://github.com/docker/compose/issues/9578
Resolves https://github.com/docker/compose/issues/9468
Resolves https://github.com/docker/compose/issues/9683

Depends on https://github.com/compose-spec/compose-go/pull/280

Documentation update [PR](https://github.com/docker/docker.github.io/pull/15161)

**(not mandatory) A picture of a cute animal, if possible in relation with what you did**
beautiful-goat

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -151,3 +151,5 @@ replace (
 	k8s.io/apimachinery => k8s.io/apimachinery v0.22.4
 	k8s.io/client-go => k8s.io/client-go v0.22.4
 )
+
+replace github.com/compose-spec/compose-go => github.com/compose-spec/compose-go v1.2.9-0.20220712021117-dcfe0889b601 // TODO Remove and bump require section when PR is merged
```
Comment (ulyssessouza) [inline go.mod:155]: Note that it points to https://github.com/compose-spec/compose-go/pull/280

File: pkg/compose/run.go [hunk 2]
```diff
@@ -116,8 +116,13 @@ func applyRunOptions(project *types.Project, service *types.ServiceConfig, opts
...
 		projectEnv := env.Resolve(func(s string) (string, bool) {
-			if _, ok := service.Environment[s]; ok {
-				return "", false
+			cmdLineVar, ok := env[s]
+			if ok && cmdLineVar == nil {
+				v, ok := project.Environment[s]
+				return v, ok
+			}
+			if sEnv, ok := service.Environment[s]; ok && sEnv != nil {
+				return *sEnv, ok
```
Comment (milas) [inline pkg/compose/run.go:125]: I'm having a hard time following this.

I _THINK_ it should be simplified/shortened to:
```go
v, ok := project.Environment[s]
return v, ok
```

My understanding is that `env` is the `cmdEnv`, and `Resolve` will call your `lookupFn` for any values that have `nil`, right?

So the immediate lookup from from `env` (aka `cmdEnv`) is guaranteed to be `nil`.
As a result, we'll always fall back to returning whatever is in the project (OS/host) env and never attempt to read from the service env here...which I think is what we want!
User did `compose run --env FOO` which explicitly means "read this value from my OS env".

If I'm missing some nuance here, some comments about the why/how you end up in each of these cases and how that relates to precedence would be super helpful.

TL;DR I think this code is _technically_ correct, but very roundabout?

File: pkg/e2e/compose_environment_test.go [hunk 3]
```diff
@@ -27,12 +27,39 @@ import (
...
-			"--project-directory", projectDir, "up", "-d", "--build")
+			"up", "-d", "--build")
+	})
+
+	// Full options activated
+	// 1. Compose file
+	// 2. Shell environment variables <-- Result expected
+	// 3. Environment file
+	// 4. Dockerfile
+	// 5. Variable is not defined
```
Comment (milas) [inline pkg/e2e/compose_environment_test.go:41]: This is great 😻

File: cmd/compose/compose.go [hunk 4]
```diff
@@ -354,10 +354,8 @@ func setEnvWithDotEnv(prjOpts *projectOptions) error {
...
 	}
 	for k, v := range envFromFile {
-		if _, ok := os.LookupEnv(k); !ok {
-			if err := os.Setenv(k, v); err != nil {
-				return err
-			}
+		if err := os.Setenv(k, v); err != nil {
+			return err
 		}
 	}
```
Comment (milas) [inline cmd/compose/compose.go:360]: Nit: Given the complexity/subtlety, I think a comment here is warranted, e.g. `// overwrite the process env with merged OS + env file results` or something

For context: I thought this was a regression at first since OS/shell has higher precedence than env file, but then I realized that `GetEnvFromFile` takes in the OS/shell env and handles the precedence/merging, which is why it's safe/necessary to overwrite everything here

**Review by milas — COMMENTED** [review]
Wahoo! Looks good for approval once the `compose-go` side of the house is merged + tagged 🥳 

(Sorry for nits - as you're well aware, the env code is pretty complex, so I was trying to identify places I got confused and had to dig in more)

File: pkg/compose/run.go (hunk 2 above)
Comment (ulyssessouza) [reply pkg/compose/run.go]: Ahhh!!! Nice catch!!!
That was an old "bad good idea" of when using `-e UNSETVAR` and it couldn't be found in `project.Environment` it could check in the `service:environment` section of the Compose file.
But... As you said, that's bad! Specially when the user imperatively used the `-e` option to make its usage explicit.

File: cmd/compose/compose.go (hunk 4 above)
Comment (ulyssessouza) [reply cmd/compose/compose.go:360]: Yep! Adding the comment.

File: pkg/compose/run.go [hunk 5]
```diff
@@ -116,8 +116,13 @@ func applyRunOptions(project *types.Project, service *types.ServiceConfig, opts
 	if len(opts.Environment) > 0 {
 		env := types.NewMappingWithEquals(opts.Environment)
 		projectEnv := env.Resolve(func(s string) (string, bool) {
```
[the remark this answers is not in the document]
Comment (ulyssessouza) [reply pkg/compose/run.go]: Sure! I felt the same while reading. Didn't change cuz it wasn't part of the PR.
Renaming in 1, 2, 3...

File: pkg/e2e/compose_environment_test.go [hunk 6]
```diff
@@ -27,12 +27,39 @@ import (
...
+	t.Run("compose file priority", func(t *testing.T) {
+		cmd := c.NewDockerComposeCmd(t, "-f", "./fixtures/environment/env-priority/compose-with-env.yaml",
+			"--env-file", "./fixtures/environment/env-priority/.env.override",
+			"run", "--rm", "-e", "WHEREAMI", "env-compose-priority")
```
Comment (nicksieger) [inline pkg/e2e/compose_environment_test.go:45]: So there's a difference between `-e WHEREAMI` with the var defined, and `-e WHEREAMI=shell`? In this case shouldn't `3.` be the expected result?

File: pkg/e2e/compose_environment_test.go [hunk 7]
```diff
@@ -43,121 +70,179 @@ func TestEnvPriority(t *testing.T) {
 	// 5. Variable is not defined
 	t.Run("compose file priority", func(t *testing.T) {
 		cmd := c.NewDockerComposeCmd(t, "-f", "./fixtures/environment/env-priority/compose-with-env.yaml",
-			"--project-directory", projectDir, "--env-file", "./fixtures/environment/env-priority/.env.override", "run",
+			"--env-file", "./fixtures/environment/env-priority/.env.override", "run",
 			"--rm", "-e", "WHEREAMI", "env-compose-priority")
 		cmd.Env = append(cmd.Env, "WHEREAMI=shell")
 		res := icmd.RunCmd(cmd)
-		assert.Equal(t, strings.TrimSpace(res.Stdout()), "Compose File")
+		assert.Equal(t, strings.TrimSpace(res.Stdout()), "override")
```
Comment (nicksieger) [inline pkg/e2e/compose_environment_test.go:77]: Update to `3. Environment file <-- Result expected` above?

Comment (nicksieger) [reply pkg/e2e/compose_environment_test.go:77]: Also, these tests so far seem to indicate that `1. Compose file` is not the highest priority. Am I reading this right?

**Review by nicksieger — COMMENTED** [review]
Overall this looks good, just have some questions about the tests and what the new priority order is.

File: pkg/e2e/compose_environment_test.go (hunk 6 above)
Comment (ulyssessouza) [reply pkg/e2e/compose_environment_test.go:45]: In the first, the variable value will be inherited from the environment. 
The second one will have the explicit value passed (`shell` in this case).

**Review by nicksieger — APPROVED** [review]

**Review by glours — APPROVED** [review]
LGTM

**Comment by K0Te** [comment]
Hi,

Sorry if this is not the correct place to ask a question, I've checked this PR source code and documentation update and am still not sure what the expected behavior is.
I've used `.env` file to override `COMPOSE_FILE` for local development: `COMPOSE_FILE=docker/docker-compose.yml:docker/docker-compose.override.yml`. This allowed to have all docker stuff in a separate folder and provide some secrets in `docker-compose.override.yml`.
And on CI I've just used `COMPOSE_FILE=docker/docker-compose.yml` as an env variable in GH Actions.

This worked fine as of docker compose 2.7.0, but with docker compose 2.9.0 this workflow doesn't work anymore. Value from `.env` started to take precedence over the environment variable.

Is this a new expected behaviour or a bug?
